include/mbgl/style/expression/expression.hpp

```diff
@@ -0,0 +1,270 @@
+#pragma once
+
+#include <array>
+#include <vector>
+#include <memory>
+#include <mbgl/util/optional.hpp>
+#include <mbgl/util/color.hpp>
+#include <mbgl/style/function/type.hpp>
+#include <mbgl/util/feature.hpp>
+#include <mbgl/style/expression/parsing_context.hpp>
+#include <mbgl/style/conversion.hpp>
+
+namespace mbgl {
+
+class GeometryTileFeature;
+
+
+namespace style {
+namespace expression {
+
+using OutputValue = variant<
+    float,
+    std::string,
+    mbgl::Color,
+    std::array<float, 2>,
+    std::array<float, 4>>;
+
+struct EvaluationError {
+    std::string message;
+};
+
+struct CompileError {
+    std::string message;
+    std::string key;
+};
+
+class Expression {
+public:
+    Expression(std::string key_, type::Type type_) : key(key_), type(type_) {}
+    virtual ~Expression() {}
+    
+    virtual optional<OutputValue> evaluate(float, const GeometryTileFeature&, EvaluationError& e) const = 0;
```

Comment:
Suggest packaging the arguments to `evaluate` in a struct, so if we need to change types or add new parameters it doesn't require touching the definition of every subexpression.